### DISCUSSION TITLE
Support for 'FM30 SIYI' Module

### DISCUSSION THREAD

**Discussion by FRPProject** [OpenI6X/opentx#108] [Q&A]
First of all, thank you for your great work. I appreciate all of your hard work. Just have a question, is it possible to add **“MULTI - SIYI”** protocol to the external receiver option? I am trying to add the FM30 SIYI module, but so far no luck with options in the menu. I have connected the module to the TX and GND of my i6X and running OpenI6X v1.1.1.  I have attached the module's manual in case you'd like to have a glance at it. Thank you.

[FM30_User_Manual_en_v1.1.pdf](https://github.com/OpenI6X/opentx/files/7808240/FM30_User_Manual_en_v1.1.pdf)

**Comment by marianomd** [comment]
Hello, did you know can flash ExpressLRS on that module?

**Reply by FRPProject** [reply]
Hi there, 

Thanks for replying to my question. 

I read about that, but flashing ExpressLRS on FM30 is a one-way process and there is no returning to the stock firmware afterward. I have great experience with the SIYI firmware. This module comes with Bluetooth connectivity that allows me to connect to QGroundControl wirelessly.  Currently, I have connected my ix6 to a Flysky X6B receiver which has an SBus port. From this receiver, I route SBus to the FM30 module to connect to the FR Mini receiver mounted on my Quad/Plane. 

This way I managed to make it work. However, in the manual of FM30, I found the following guide claiming that it supports OpenTx. This is why I asked if it is possible to add “MULTI - SIYI” on the Open-i6x firmware you are developing.

![image](https://user-images.githubusercontent.com/37748115/148122312-abf8d1ca-a11c-4ccd-a8bf-9269b89176fb.png)

Again, thanks for your fantastic development.

**Reply by marianomd** [reply]
Thanks! I see that it also supports a PPM mode, you can connect to it that way too. Regarding "MULTI", I have no idea. I don't know if that is related to the "4in1 multi protocol module". We don't have none of those modules to test, so this feature will need a dev that has one and wants to implement it.

**Reply by FRPProject** [reply]
PPM is limited to 8 channels while S.Bus offers 16 channels. I can test out any implementation that supports FM30 and provide feedback. Never been a dev, but I'd like to help out if I can. Since 'MULTI' is already in the openTX firmware, I was hoping maybe there is a way to copy their implementation to open-i6x.

**Reply by ajjjjjjjj** [reply]
> This module comes with Bluetooth connectivity that allows me to connect to QGroundControl wirelessly.

ExpressLRS also support that on this module, it's called 'BT Telemetry' and supported on OpenI6X.

**Reply by FRPProject** [reply]
Thanks for following up. I will look into that; Still, it would be nice to add the multi protocol if it is possible. Many people are also using it for other multi protocol modules: https://github.com/pascallanger/DIY-Multiprotocol-TX-Module

**Reply by FRPProject** [reply]
I finally decided to flash ExpressLRS on the FM30 module. I have installed the newest version available which is 2.4.0 and also upgraded the FS-i6X to the 1.4 firmware. The FM30 gets 5V power from the transmitter, TX2 is connected to the signal pin on FM30, GND to the FM30's GND pin. I have turned on the external module in the transmitter settings. 

Unfortunately, FS-i6X ExpressLRS doesn't realize that the module is connected. It just displays "Loading ?/????". I tried connecting the TX2 pin to the FM30's S.Port pin but nothing happened. Then I flashed older ExpressLRS on the FM30 module but still, the transmitter doesn't recognize the module.

Any idea what may resolve this connectivity issue? 
What should I do with the LUA script that the ExpressLRS configurator generates?

Thanks a lot.

**Reply by ajjjjjjjj** [reply]
Enable external module CRSF protocol and try again.

**Reply by FRPProject** [reply]
I have done that, still the same issue.

**Reply by FRPProject** [reply]
I also flashed the FS-i6X with the 1.3 firmware but again the same problem.

**Reply by ajjjjjjjj** [reply]
Maybe try to lower Max bauds in hardware tab? Wild guess.

**Reply by FRPProject** [reply]
I just tested on all different max bauds, didn't helped though. Any other ideas? I am open to any other suggestions at this point.

**Reply by ajjjjjjjj** [reply]
I've just realized that max bauds is ignored on all releases, can you try on latest master?
https://app.circleci.com/pipelines/github/OpenI6X/opentx/470/workflows/739c0720-1b6c-4fe4-bd61-2446a4721ec5/jobs/876/artifacts

You may also check all your connections again.
Is it working on any other radio?

**Reply by FRPProject** [reply]
Well this is my main radio, I followed this 
https://www.rcgroups.com/forums/showatt.php?attachmentid=14899887&d=1619245493
and everything seems okay. I have checked the wires multiple times, they are connected physically.

I am gonna flash the radio and get back to you in a few minutes with those. Thanks.

**Reply by FRPProject** [reply]
Just finished testing with the new firmware (firmware_i6x-elrsv2_en_220417.bin); turned the CRSF setting ON. Each time, I waited for the square indicator to disappear before testing the new baud rate. Still, it says loading, no connectivity to the radio.

**Reply by FRPProject** [reply]
not sure if this is gonna help or not but this is the LUA script from the configurator that I was supposed to upload to the SD card according to [the tutorial I followed to flash the FM30 module](https://www.rcgroups.com/forums/showatt.php?attachmentid=14899887&d=1619245493). This is the only part I could not follow.

[elrsV2.lua.txt](https://github.com/OpenI6X/opentx/files/8502245/elrsV2.lua.txt)

**Reply by ajjjjjjjj** [reply]
You need to wait for the square icon only before shutdown, other than that change is immediate.
Try to reflash your module or try to find someone with radio that will allow to confirm that it is not the TX module.
LUA is the same for all, so this is not the problem.

**Reply by FRPProject** [reply]
Thanks for your feedback. 
I am gonna flash it to expressLRS V1.0.0, maybe it helps. Unfortunately, I don't own another radio or have access to others to borrow one at the moment.

**Reply by ajjjjjjjj** [reply]
Don't go any lower than 2.0. That's the lowest supported version.

**Reply by FRPProject** [reply]
Oh, I see. I'll go for V2.0.0 then. Thanks.

**Reply by FRPProject** [reply]
I tested expressLRS V2.0, V2.2, and V2.4, none of them worked. I changed the baud rates as well and made sure the CRSF external module setting is ON. This is very unfortunate. I cannot even return to the stock SIYI firmware.

**Reply by FRPProject** [reply]
**Update:** I have also flashed my FR Mini to be expresslrs TX (V2.4.0) based on [this tutorial](https://www.expresslrs.org/2.0/quick-start/transmitters/tx-siyifm30/#fr-mini-rx-as-tx) and it was successful. The LED lights up as intended. But again I have the same connectivity issue between the i6x-OpenTX and the expresslrs module. I have checked the physical connections, no problem there. 

**What could be possibly wrong?** @marianomd  @ajjjjjjjj 

Is there any chance the SIYI LUA script contains some unique connectivity inputs compared to the other modules? Please see the file attached.  Thanks.
[elrsV2_SIYIFM30.lua.txt](https://github.com/OpenI6X/opentx/files/8508728/elrsV2_SIYIFM30.lua.txt)

**Reply by ajjjjjjjj** [reply]
Try to provide better 5V source, maybe the one from radio do not provide enough current causing TX to constantly reset.

**Reply by FRPProject** [reply]
I tried again with 3S Lipo to 5V BEC, and powered the module directly from the BEC, still says loading even at different baud rates. I also made a common ground between the radio's GND3 Pin and the module's GND, but didn't help.

**Reply by FRPProject** [reply]
I am not sure if this means anything but I was trying to figure out if the pins are producing any sort of signals or not using Arduino analog pin reader. It seems they are alive:
![signals_arduinoOSci](https://user-images.githubusercontent.com/37748115/164095001-98d8e969-55c3-4cb2-b393-176f093040b6.png)

Does this give you any hint as to what is going on? Thanks.

**Reply by ajjjjjjjj** [reply]
What the number on vertical scale represents? Because on FM30 it looks quite different than on i6X and FR-Mini.

**Reply by FRPProject** [reply]
Thanks for getting back to me. 

I followed [this tutorial](https://www.youtube.com/watch?v=uYnaEBqN0Z4) to make a simple Arduino-based oscilloscope. I think the vertical axis should be a measure of voltage (probably mV) as I used [analogRead()](https://www.arduino.cc/en/Reference/AnalogRead). 

Now, the reason they have different signal patterns is maybe that when I flashed FM30 with ELRS, I selected the 10mW option in the configurator than the normal one that I chose for FR-the Mini. Please note that I have to use 3.3V to power up the FR-Mini and 5V for FM30.

**Reply by ajjjjjjjj** [reply]
That setting does not affect communication. For me it looks like something is wrong with that output, signal is never reaching gnd, like missing pull down or non common gnd? You can try ask on ExpressLRS project issues page. There is more people there, maybe someone have seen this kind of problem before.

**Reply by FRPProject** [reply]
Thanks, so you think my radio is functioning as intended on TX2 pin? I thought maybe there was something wrong with my radio's hardware. 

I see what you mean but for FR-Mini, the signal seems to reach the GND level, doesn't it? still, no connectivity to the radio. Do you have a signal pattern for your expressLRS modules that you use? 

I will try to reach out to the ELRS community as well. Hopefully, I find a solution for this issue.

**Reply by FRPProject** [reply]
Just got into another issue!

After all of that effort to connect FM30 ELRS, I decided to bind to the normal flysky receivers (iA6B, and X6B). I cannot bind my radio to them at all, even I flashed the original flysky firmware to the radio, still cant bind! It seems somehow I managed to damage the hardware while I was trying to get FM30 to work.

**Reply by ajjjjjjjj** [reply]
Were you GND between radio and TX always connected? If no, that's possible.
Sorry to hear, we ave a one thread in issues with similar conclusion - broken mcu.

**Reply by FRPProject** [reply]
Well, I did not really know that GND between radio and the external TX module could cause such a permanent hardware damage. The radio flashes fine and interface is all good, but from the hardware perspective, it seems something is wrong for sure. Sadly, I have to say goodbye to my old trustee Flysky i6x radio.

**Reply by FRPProject** [reply]
Got a brand new Flysky i6x radio. I had to flash the firmware through [QMK Toolbox](https://github.com/qmk/qmk_toolbox) as the new board comes with APM32 DFU ISP and it is no longer STM32 (also see [here](https://www.rcgroups.com/forums/showthread.php?3916435-FlySky-I6X-port-of-OpenTX/page15)). 

I tested my FM30 (flashed with ELRS 2.4) again but still no luck. I am waiting for an ELRS module (Jumper 2.4G ELRS TX Module AION NANO) to see if it works. 

By the way, thanks for the new firmware update.

**Reply by ajjjjjjjj** [reply]
OK, looking forward to your feedback with new module.

**Reply by FRPProject** [reply]
So I got a HappyModel ELRS ES24TX TX Module, ELRS works great even on the older i6x radio that I thought was broken :) I guess there is something wrong with the FM30 module. I did everything I possibly could and had a very hard time with this module. 

I meant to ask you about PC13 pin on i6x, the picture says it is an external power switch. How should I use it? Could you please elaborate. Thank a lot.

**Reply by ajjjjjjjj** [reply]
Of course, i'm happy to see it finally working for you!
External module power switch: on v1.5.0 and previous versions it is `PC9`, next release and current master have it changed to `PC13`.
This pad is high (3.3V) when external module should be on, and low (0V) when not. This can be used to switch i.e. mosfet driving module power or can be connected directly to power converter `EN` (enable) pin if available. I don't want to explain it any further, basically it needs some electronics knowledge to do it right.

**Reply by FRPProject** [reply]
I see, thanks a lot. I'll look into that.

**Reply by Spacekip** [reply]
I've just stumbled upon this thread, trying to get a FM30 working with my I6X as well.
After reading everything, I guess there's no way to get this working really :(

Or did you finally manage to get it working FRPProject?

**Reply by FRPProject** [reply]
@Spacekip If you want to use FM30 as is, perhaps the safest way is to use the PPM pins (https://www.youtube.com/watch?v=m4Gc57gW3oQ&t=499s)  or (https://www.youtube.com/watch?v=RDeRIGQFtEE&t=188s). I tried to flash my FM30 TX and its receivers with ELRS but every attempt I did was failed. I think it was a matter of proper ELRS flashing. SIYI has an additional uncommon requirement on flashing its hardware which needs an RST pin. I was able to unlock and remove the original SIYI hardware firmware, but then I couldn't flash ELRS properly after many many attempts. I ended up buying an ELRS module with its receivers.
